### ISSUE TITLE
RecursionError in some sort of obscure IPython/prettyprinter interaction case

### ISSUE THREAD

**Issue by njsmith** [tommikaikkonen/prettyprinter#3]
* PrettyPrinter version: 0.4.0
* Python version: 3.6
* Operating System: Linux

After configuring IPython to use `prettyprinter` by putting this in my IPython startup file:

<details>

```python
try:
    import prettyprinter
except ImportError:
    pass
else:
    from prettyprinter.color import GitHubLightStyle
    ipy = get_ipython()
    ipy.colors = 'LightBG'
    ipy.highlighting_style = GitHubLightStyle

    prettyprinter.install_extras(
        # Comment out any packages you are not using.
        include=[
            'ipython',
            'attrs',
            #'django',
        ],
        warn_on_error=True
    )
```
</details>

I was then curious how `prettyprinter` would work with code that actually tries to take advantage of IPython's smart pretty printer. (To reproduce, I think an environment with `pip install ipython patsy prettyprinter` should be sufficient.) The result was, well... not pretty:

```
In [1]: import patsy

In [2]: patsy.parse_formula.parse_formula("y ~ x1 + x2")
Out[2]: ---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
~/.user-python3.6-64bit/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/extras/ipython.py in pretty(self, obj)
     25                 style = None
     26 
---> 27             cpprint(obj, stream=self.stream, style=style, width=columns, end=None)
     28 
     29         def flush(self):

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/__init__.py in cpprint(object, stream, indent, width, depth, compact, ribbon_width, style, end)
    133         width=width,
    134         depth=depth,
--> 135         ribbon_width=ribbon_width
    136     )
    137     if stream is None:

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/prettyprinter.py in python_to_sdocs(value, indent, width, depth, ribbon_width)
   1406     doc = pretty_python_value(
   1407         value,
-> 1408         ctx=PrettyContext(indent=indent, depth_left=depth, visited=set())
   1409     )
   1410 

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/prettyprinter.py in pretty_python_value(value, ctx)
    378         doc = pretty_dispatch(
    379             value,
--> 380             ctx
    381         )
    382 

~/.user-python3.6-64bit/lib/python3.6/functools.py in wrapper(*args, **kw)
    801 
    802     def wrapper(*args, **kw):
--> 803         return dispatch(args[0].__class__)(*args, **kw)
    804 
    805     registry[object] = func

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/prettyprinter.py in _run_pretty(pretty_fn, value, ctx, trailing_comment)
    329                 raise e
    330     else:
--> 331         doc = pretty_fn(value, ctx)
    332 
    333     if not (

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/prettyprinter.py in _repr_pretty(value, ctx)
    357         if predicate(value):
    358             return fn(value, ctx)
--> 359     return repr(value)
    360 
    361 

~/.user-python3.6-64bit/lib/python3.6/site-packages/patsy/util.py in repr_pretty_delegate(obj)
    420     if optional_dep_ok and "IPython" in sys.modules:
    421         from IPython.lib.pretty import pretty
--> 422         return pretty(obj)
    423     else:
    424         return _mini_pretty(obj)

~/.user-python3.6-64bit/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(obj, verbose, max_width, newline, max_seq_length)
    138     stream = StringIO()
    139     printer = RepresentationPrinter(stream, verbose, max_width, newline, max_seq_length=max_seq_length)
--> 140     printer.pretty(obj)
    141     printer.flush()
    142     return stream.getvalue()

... last 9 frames repeated, from the frame below ...

~/.user-python3.6-64bit/lib/python3.6/site-packages/prettyprinter/extras/ipython.py in pretty(self, obj)
     25                 style = None
     26 
---> 27             cpprint(obj, stream=self.stream, style=style, width=columns, end=None)
     28 
     29         def flush(self):

RecursionError: maximum recursion depth exceeded while calling a Python object
```

This is trying to pretty-print a [`patsy.infix_parser.ParseNode`](https://github.com/pydata/patsy/blob/924b387652d13f816fd4a32983aae2ccba38fd84/patsy/infix_parser.py#L75), whose [`__repr__`](https://github.com/pydata/patsy/blob/924b387652d13f816fd4a32983aae2ccba38fd84/patsy/infix_parser.py#L82) uses a helper [`repr_pretty_delegate`](https://github.com/pydata/patsy/blob/924b387652d13f816fd4a32983aae2ccba38fd84/patsy/util.py#L432-L433) which just calls `IPython.lib.pretty.pretty`, because I didn't want to reimplement `__repr__` twice, once normally and once using IPython's system. I guess the problem is that you're monkeypatching `IPython.lib.pretty.RepresentationPrinter.pretty`, so we end up in a loop where you don't recognize my object so you try to call `__repr__`, then my `__repr__` tries to call IPython's `pretty`, which goes back into your code and doesn't recognize my object so you try to call `__repr__`...

(BTW, you may also appreciate [`repr_pretty_impl`](https://github.com/pydata/patsy/blob/924b387652d13f816fd4a32983aae2ccba38fd84/patsy/util.py#L437-L454), which is exactly the same interface as your `pretty_call`, but implemented using IPython's pretty-printer and about 5 years ago. And in case you're curious, here's what IPython prints for that code if you don't have prettyprinter enabled:
<details>

```
In [2]: patsy.parse_formula.parse_formula("y ~ x1 + x2")
Out[2]: 
ParseNode('~',
          Token('~', <Origin y ->~<- x1 + x2 (2-3)>),
          [ParseNode('PYTHON_EXPR',
                     Token('PYTHON_EXPR',
                           <Origin ->y<- ~ x1 + x2 (0-1)>,
                           extra='y'),
                     []),
           ParseNode('+',
                     Token('+', <Origin y ~ x1 ->+<- x2 (7-8)>),
                     [ParseNode('PYTHON_EXPR',
                                Token('PYTHON_EXPR',
                                      <Origin y ~ ->x1<- + x2 (4-6)>,
                                      extra='x1'),
                                []),
                      ParseNode('PYTHON_EXPR',
                                Token('PYTHON_EXPR',
                                      <Origin y ~ x1 + ->x2<- (9-11)>,
                                      extra='x2'),
                                [])])])
```
</details>
So you can see why I'm interested in this... :-))

Any ideas on how to avoid this? I suppose one obvious solution would be to go through and register `prettyprinter` printers for all my objects. It would be nice if `prettyprinter` could do something more out-of-the-box compatible here, though, so to improve the possibilities of it becoming a standard thing in IPython -- maybe if an object has a `_repr_pretty_` it could dispatch back into the real un-monkeypatched IPython formatting system?

Also, is there any way to do something like this `__repr__` → `prettyprinter` trick to avoid implementing all my reprs multiple times?

**Comment by tommikaikkonen** [comment]
Thanks for opening the issue! Seems like a valid use case `prettyprinter` should support.

Going to think about this more, but for a temporary workaround, you may be able to delegate to `__repr_pretty__` the same way I check for `__attrs_attrs__` to identify classes created with the `attrs` library:

https://github.com/tommikaikkonen/prettyprinter/blob/master/prettyprinter/extras/attrs.py#L46

then, in the pretty printer function you register with `register_pretty`, you would return the string representation from the not-monkeypatched `IPython.lib.pretty`. It won't be syntax-highlighted but it shouldn't crash.

**Comment by tommikaikkonen** [comment]
I spent some time a week ago trying to find a cleaner way to integrate with IPython, but I couldn't find a simple approach. It seemed like I would have had to override a lot more classes to get around monkeypatching, and I'm not familiar enough with the IPython codebase to start doing that :) If you or anyone else has ideas for a cleaner integration, let me know!

I previously glossed over your comment about filling in `__repr__` using `prettyprinter`. I think that would be worthwhile to add to the library. I encountered the same problem a couple days back, using `pdb` to look at values, which doesn't have `prettyprinter` integrated and therefore shows the uninformative default repr. I think an interface like

```python
from prettyprinter import pretty_repr

class MyClass:
    __repr__ = pretty_repr

@register_pretty(MyClass)
def pretty_myclass(value, ctx):
    # pretty_repr will dispatch to here
```

would work. `pretty_repr` would call `pformat` with some parameters optimized for reprs.

Re: supporting `_repr_pretty_`, like I mentioned in https://github.com/ipython/ipython/issues/10953, it may be possible to support it directly... but I haven't looked into it yet. Did you have any luck with a workaround? Will gladly accept patches to the IPython integration to improve it.

(So cool that you implemented `repr_pretty_impl` way back, it's definitely an intuitive interface!)

**Comment by tommikaikkonen** [comment]
`pretty_repr` for filling in `__repr__` methods was added in 0.8.0 https://prettyprinter.readthedocs.io/en/latest/api_reference.html#prettyprinter.pretty_repr

PR #6 implements support for the `_repr_pretty_` interface used by `IPython.lib.pretty`, which provides the following output now:

<img width="978" alt="repr_pretty_patsy" src="https://user-images.githubusercontent.com/2487359/34472625-dc8c669c-ef1a-11e7-994e-f5ef464f6008.png">

Recursive calls to `p.pretty` in `_repr_pretty_` implementations are delegated to the normal `prettyprinter`, which is able to produce colorized output seen in the above screenshot.

**Comment by njsmith** [comment]
That's *gorgeous*.

**Comment by tommikaikkonen** [comment]
Fixed in 0.9.0 by installing the `'ipython_repr_pretty'` extra:

```python
from prettyprinter import install_extras
install_extras(['ipython_repr_pretty'])
```
